[ci] Check contrib modules are using the correct core collector version

hickeyma opened the pull request:
Description from @mx-psi in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/8546#issuecomment-1084592613:

> My intention is to avoid the following situation:
> 1. Someone creates a PR adding a new examplemodule module, targeting Collector core vN module
> 2. Weeks pass, it gets reviewed and merged just before the vN+3 contrib release, still targeting vN
> 3. The release manager for vN+3 has then to manually update the examplemodule module code to be compatible with vN+3
>
> I want to avoid (3) and, instead, make the contributor adding examplemodule do that job before merging the PR, by comparing the Collector core version at the top-level go.mod file with the Collector core version at the examplemodule go.mod file.

**The result of check-collector-module-version (no incorrect versions):**

```console
There are 0 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
There are 0 incorrect "go.opentelemetry.io/collector/model" version(s) in the module files.
```

**The result of check-collector-module-version (incorrect versions):**

```console
Incorrect version "v0.47.0" of "go.opentelemetry.io/collector " is included in "./receiver/couchbasereceiver/go.mod". It should be version "v0.48.0".
Incorrect version "v0.44.0" of "go.opentelemetry.io/collector " is included in "./exporter/awsemfexporter/go.mod". It should be version "v0.48.0".
There are 2 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
```

**The result of check-collector-module-version (incorrect model versions):**

```console
There are 0 incorrect "go.opentelemetry.io/collector " version(s) in the module files.
Incorrect version "v0.43.0" of "go.opentelemetry.io/collector/model" is included in "./exporter/datadogexporter/go.mod". It should be version "v0.48.0".
There are 1 incorrect "go.opentelemetry.io/collector/model" version(s) in the module files.
```

Fixes #8546 

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Check contrib modules are using the correct core collector version

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

mx-psi commented on .github/workflows/scripts/check-collector-module-version.sh:
```diff
@@ -0,0 +1,53 @@
+get_collector_version() {
```
Can you add a shebang?
```suggestion
#! /bin/bash

get_collector_version() {
```

Also, should we `set -euxo pipefail`?

hickeyma replied:
Sorry, forgot to round off the file before pushing. Done.

mx-psi commented on .github/workflows/scripts/check-collector-module-version.sh:
```diff
@@ -0,0 +1,53 @@
+COLLECTOR_MODULE="go.opentelemetry.io/collector "
+COLLECTOR_MODEL_MODULE="go.opentelemetry.io/collector/model"
+MAIN_MOD_FILE="./go.mod"
+COLLECTOR_MOD_VERSION=$(get_collector_version $COLLECTOR_MODULE $MAIN_MOD_FILE)
```
Shellcheck suggests doing:

```suggestion
COLLECTOR_MOD_VERSION=$(get_collector_version "$COLLECTOR_MODULE" "$MAIN_MOD_FILE")
```

for good argument hygiene

hickeyma replied:
Agree, done.

mx-psi commented on .github/workflows/scripts/check-collector-module-version.sh:
```diff
@@ -0,0 +1,53 @@
+   fi
+}
+
+COLLECTOR_MODULE="go.opentelemetry.io/collector "
+COLLECTOR_MODEL_MODULE="go.opentelemetry.io/collector/model"
```
I guess this works because you explicitly add a space after `COLLECTOR_MODULE`, is that right? If so, I would rather do `grep "$collector_module " "$mod_file"` and add a comment on the grep line documenting what the space is for.

hickeyma replied:
That is probably a better place to put it. However, `check_collector_version_correct` is called using the `COLLECTOR_MODULE` and `COLLECTOR_MODEL_MODULE` strings. `COLLECTOR_MODEL_MODULE` does NOT have a space at end. For this reason, I think it will need to remain here in the initialisation of the string. I have added a comment. Let me know if that is ok.

hickeyma committed:
Updates after review

Review:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9112#pullrequestreview-935066186

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for the review @mx-psi. dated and ready for review again.

mx-psi reviewed (approved):
Tested locally, seems to work fine. I would like another approver/maintainer to take a look before I mark as ready to merge

mx-psi commented:
And thanks for doing this @hickeyma 🙇 I think this will make releases a bit smoother

jpkrohling commented on .github/workflows/scripts/check-collector-module-version.sh:
```diff
@@ -0,0 +1,82 @@
+   fi
+}
+
+# Compare the collecor main core version against all the collector component
```
```suggestion
# Compare the collector main core version against all the collector component
```

hickeyma replied:
Done

jpkrohling commented on .github/workflows/scripts/check-collector-module-version.sh:
```diff
@@ -0,0 +1,82 @@
+
+# Compare the collecor main core version against all the collector component
+# modules to verify that they are using this version as its dependency
+check_collector_version_correct() {
```
This is checking all go.mod files against the main go.mod's, right? Perhaps using "versions" (plural) would be more appropriate for the function name?

```suggestion
check_collector_versions_correct() {
```

hickeyma replied:
That makes sense. Updated.

jpkrohling reviewed (commented):
LGTM, just a couple of minor comments. Would you be able to add example outputs as a comment to this PR?

hickeyma committed:
Update after review

Review:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9112#pullrequestreview-936398313

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks for the review and feedback @jpkrohling. I have updated with your comments. It is ready for review again.

> Would you be able to add example outputs as a comment to this PR?

I added output to the description.
